packages/scrapers-v2/src/lexer/lexer.ts

```diff
@@ -0,0 +1,432 @@
+    yearVersion,
+    majorName,
+    sections: await sections,
+  };
+};
+
+/**
+ * Produces overall HSections for each HTML table in the page
+ * @param $
+ * @param requirementsContainer
+ */
+const tokenizeSections = async (
+  $: CheerioStatic,
+  requirementsContainer: Cheerio
+): Promise<HSection[]> => {
+  // use a stack to keep track of the course list title and description
+  const descriptions: string[] = [];
+  const courseList: HSection[] = [];
+
+  for (const element of requirementsContainer.children().toArray()) {
+    if (HEADER_REGEX.test(element.name)) {
+      descriptions.push(parseText($(element)));
+    } else if (
+      element.name === "table" &&
+      element.attribs["class"] === "sc_courselist"
+    ) {
+      const tableDesc = descriptions.pop() || "";
+      const courseTable = {
+        description: tableDesc,
+        entries: tokenizeRows($, element),
+      };
+      courseList.push(courseTable);
+    } else if (
+      // only necessary for business concentrations
+      element.name === "ul" &&
+      parseText($(element).prev()).includes("concentration")
+    ) {
+      // parse all the business concentration links
+      const links = constructNestedLinks($, element);
+      const pages = await Promise.all(links.map(loadCatalogHTML));
+      const containerId = "#concentrationrequirementstextcontainer";
+      const concentrations = await Promise.all(
+        pages.map((concentrationPage) =>
+          tokenizeSections(concentrationPage, concentrationPage(containerId))
+        )
+      );
+      courseList.push(...concentrations.flat());
+    }
+  }
+
+  return courseList;
+};
+
+/**
+ * Finds and fetches nested links, for majors with concentration requirements
+ * on separate pages.
+ * @param $
+ * @param element
+ */
+const constructNestedLinks = ($: CheerioStatic, element: CheerioElement) => {
+  // TODO: add support to non-current catalogs
+  const base = "https://catalog.northeastern.edu";
+  const concentrationsTag = "#concentrationrequirementstext";
+  return $(element)
+    .find("li > a")
+    .toArray()
+    .map((link) => $(link).attr("href"))
+    .map((link) => {
+      const url = new URL(link, base);
+      url.hash = concentrationsTag;
+      return url.toString();
+    });
+};
+
+/**
+ * Converts tables rows into a list of HRows
+ * @param $
+ * @param table
+ */
+const tokenizeRows = ($: CheerioStatic, table: CheerioElement): HRow[] => {
+  const courseTable: HRow[] = [];
+
+  for (const tr of $(table).find("tbody > tr").toArray()) {
+    // different row type
+    const type = getRowType($, tr);
+    const row = constructRow($, tr, type);
+    courseTable.push(row);
+  }
+
+  return courseTable;
+};
+
+/**
+ * Pre-parses the row to determine its type
+ * @param $
+ * @param tr
+ */
+const getRowType = ($: CheerioStatic, tr: CheerioElement) => {
+  const trClass = tr.attribs["class"];
+  const td = $(tr.children[0]);
+  const tdClass = $(tr).children()[0].attribs["class"];
```

Comment:
ah so when it's being wrapped around with $, it gives you a Cheerio. And since attribs is a function for CheerioElement, doing `$(tr).children()[0]` will give you a child which is a CheerioElement. But good point since it can also be written as `$(tr.children)[0]`

Reply:
ahhhhhh i see that makes it much clearer for me-- thank you both for clarifying!!